Flexible Page Basic Structure

KludgeKML opened the pull request:
, [Jira issue PNP-6387](https://gov-uk.atlassian.net/browse/PNP-6387)⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

## What

Adds basic support for Flexible Pages, and two section types (page_title and rich_content), suitable for making a basic version of the existing history pages. Also adds an example history page loaded from local content at test/flexible-page

(Note you will need [ALLOW_LOCAL_CONTENT_ITEM_OVERRIDE=true](https://github.com/alphagov/frontend/blob/main/docs/local-content-items.md) set to see this item - I have set it on the preview app).

Example on preview app:
https://govuk-frontend-app-pr-4718.herokuapp.com/test/flexible-page

## Why

https://github.com/alphagov/govuk-rfcs/pull/182

Trello card: https://trello.com/c/Bwy1WiJW

KludgeKML committed:
Add basic route/controller/view for Flexible Page

KludgeKML committed:
Add empty flexible_page model

KludgeKML committed:
Add PageTitle flexible section

- This flexible section renders a page title with an optional
  context, using the heading component.

Co-authored-by: Andy Sellick <andy.sellick@digital.cabinet-office.gov.uk>

KludgeKML committed:
Add RichContent flexible section

- This flexible section renders a 1/3rd 2/3rd section of page,
  in which the left (1/3) side is a contents list for the right
  (2/3) section, which is govspeak.
- Also add a contents list presenter that will return the content
  list in a format required by the contents list component.

Co-authored-by: Andy Sellick <andy.sellick@digital.cabinet-office.gov.uk>

KludgeKML committed:
Add request spec

- At this point we have a controller, a route, and a view,
  so we can add the basic tests we're currently doing as
  review tests (basic 200, and correct template rendered)

KludgeKML committed:
Add system test

- System tests for all flexible page sections to be added here
  (it may be that we make those effectively individual view tests
  later, similar to components, but for the moment it's simplest
  to add here)

KludgeKML committed:
Add FlexibleSection attribute to the FlexiblePage model and view

- We add a loop in the view to find renderable views for the sections
  and render them in place.

KludgeKML committed:
Add breadcrumbs

- Here we're relying on the contextual breadcrumb component to
  make a breadcrumb using the links/parent elements in the content
  item, but because of the full-width layout we have to opt out of
  the breadcrumb code in the layout and explicitly render it inside
  a govuk-width container.

deborahchua commented on spec/models/flexible_page/flexible_section/rich_content_spec.rb:
```diff
@@ -0,0 +1,19 @@
+RSpec.describe FlexiblePage::FlexibleSection::RichContent do
+  subject(:page_title) do
```
Should this be `:rich_content`?

leenagupte replied:
Good spot Deb!

KludgeKML replied:
Fixed!

deborahchua reviewed (commented):
Looks good to me! I've left one small comment

leenagupte commented on app/models/contents_list.rb:
```diff
@@ -0,0 +1,7 @@
+class ContentsList
```
I think perhaps this should renamed as "ContentsList" implies a specific rendering choice, and this is really just a list of links. That should mean that it'll be more likely to be reused.

KludgeKML replied:
This has been replaced with the existing ContentsOutline

leenagupte commented on app/models/flexible_page/flexible_section/base.rb:
```diff
@@ -0,0 +1,13 @@
+class FlexiblePage
+  module FlexibleSection
```
I've never seen a module nested inside a class before. It looks a bit odd.

KludgeKML replied:
I thought we were confined by rubocop to explicit nesting (in which case this is the only way we can do this), but I checked in landing pages and we were doing `module LandingPage::Block`, so I've used that style now.

leenagupte commented on app/models/flexible_page/flexible_section/rich_content.rb:
```diff
@@ -0,0 +1,15 @@
+
+        @contents_list = ContentsList.new(data["contents_list"])
+        @govspeak = data["govspeak"]
+        @image = data["image"]
```
I'd be nice to model images too.

KludgeKML replied:
Have done a simple Data model for now.

leenagupte commented on app/presenters/contents_list_presenter.rb:
```diff
@@ -0,0 +1,9 @@
+  end
+
+  def to_component
+    @contents_list.items.map { |item| { href: item.href, text: item.text } }
```
Why does this feel like it's a duplicate of something? Perhaps the ContentsList model should have a `to_h` method that returns this?

KludgeKML replied:
(this file no longer present)

leenagupte reviewed (commented):
This is looking good. The rendered page looks great! 🎉 

I have a few inline questions and comments.

andysellick commented on app/views/flexible_page/flexible_sections/_rich_content.html.erb:
```diff
@@ -0,0 +1,15 @@
+
+<div class="govuk-grid-column-two-thirds">
+  <%= render "govuk_publishing_components/components/govspeak", {} do %>
+    <%= flexible_section.govspeak.html_safe %>
```
I think initially when I wrote this it wouldn't display the govspeak properly but adding `.html_safe` fixed it, just want to check that this is the best solution for this problem.

KludgeKML replied:
Yes, we want this here. The govspeak block is HTML, putting `.html_safe` in the view here is a good way to signal "we know this is HTML we want to render, and we trust it".

andysellick commented on app/views/flexible_page/show.html.erb:
```diff
@@ -7,6 +7,20 @@
+      breadcrumbs: [
+        {
+          title: "Home",
+          url: "/",
```
For MVP breadcrumbs this won't match what it should be for history pages (Home -> History of the UK Government) so should we either hard code that in here or include it in the YAML somehow? (again to note that I'm raising a problem on code I think I wrote).

KludgeKML replied:
I've updated this to hardcode the history breadcrumb for now, since that's the scope of this scaffolding.

andysellick reviewed (commented):
Looks great overall. Just raising a few questions on some bits, primarily on code that I think I originally wrote, which seems unfair of me! But happy to have conversations around these things and figure them out.

KludgeKML committed:
Add fixture for tests

- This fixture is over-detailed at this point, but matches
  the local content item for development that will be added
  later. It's necessary to have a file to run the request
  test (it doesn't have to be this detailed, but it doesn't
  hurt). It's modelled after the No 10 history page:
  https://www.gov.uk/government/history/10-downing-street

KludgeKML commented on lib/data/local-content-items/test/flexible-page.yml:
```diff
@@ -0,0 +1,242 @@
+      <p>These Roman settlements, and those of the Anglo-Saxons and Normans who supplanted them, were not very successful. The area was prone to plague and its inhabitants were very poor. A charter granted by the Mercian King Offa in the year 785 refers to “the terrible place called Thorney Island”. It took royal patronage to give the area prestige. King Canute (reigned 1017 to 1035) built a pala ...
+
+      <figure>
+      <img src="https://www.gov.uk/assets/government-frontend/history/buildings/whitehall-072ffd42bb410697f2f1d4321126dae7a54adca7c4efed9ed6a748d5cdaf902e.jpg" alt="Whitehall from St James’s Park – Hendrick Danckerts c.1675">
```
Let's leave them as is at the moment, I think solving images in govspeak is something we have to ask publishing about.

andysellick replied:
Fair enough.

KludgeKML committed:
Add example history page

- This page modelled on the Number 10 history page
  https://www.gov.uk/government/history/10-downing-street

KludgeKML commented on app/models/link.rb:
```diff
@@ -0,0 +1,8 @@
+class Link
+  attr_reader :href, :text
+
+  def initialize(hash)
```
Have just got rid of Links for the moment, they were there to support the contents list but that's now been replaced with a ContentsOutline.

KludgeKML commented on app/views/flexible_page/flexible_sections/_rich_content.html.erb:
```diff
@@ -0,0 +1,15 @@
+<nav class="govuk-grid-column-one-third">
+  <% if flexible_section.image %>
+    <img alt="<%= flexible_section.image["alt"] %>" loading="lazy" class="govuk-!-width-full" src="<%= flexible_section.image["src"] %>">
```
Yep! Done.

leenagupte commented on spec/models/flexible_page/flexible_section/rich_content_spec.rb:
```diff
@@ -0,0 +1,19 @@
+  end
+
+  describe "#initialize" do
+    it "sets the attributes from the contents hash" do
```
Image tests should be added here too.

KludgeKML replied:
Added

leenagupte commented on spec/system/flexible_page_spec.rb:
```diff
@@ -0,0 +1,37 @@
+      visit base_path
+
+      expect(page).to have_selector("h1", text: "10 Downing Street")
+      expect(page).to have_text("History")
```
I think a test is needed for lead paragraph if it exists.

KludgeKML replied:
Added

leenagupte commented on app/helpers/application_helper.rb:
```diff
@@ -24,7 +24,7 @@ def current_path_without_query_string
     return false if content_item.nil?
 
-    no_breadcrumbs_for = %w[homepage landing_page service_manual_homepage service_manual_service_toolkit]
+    no_breadcrumbs_for = %w[flexible_page homepage landing_page service_manual_homepage service_manual_service_toolkit]
```
Out of interest what does the breadcrumb look like if you do use application helper? I wonder if that's ok, rather than having to hard-code the breadcrumb. I'd imagine you'd want to make the basic structure as generic as possible.

leenagupte replied:
I think if you add a "parent" link to the content item, it will fix the breadcrumb issue.

KludgeKML replied:
Reworked now.

leenagupte reviewed (commented):
This is looking good. I've added a few inline comments, mainly where there's a missing test.

andysellick reviewed (commented):
I won't approve for now as there's some comments still to address from other reviewers, but I'm happy with this code, good job 👏

KludgeKML commented on app/models/flexible_page/flexible_section/base.rb:
```diff
@@ -0,0 +1,11 @@
+    attr_reader :data, :flexible_page, :type
+
+    def initialize(flexible_section_hash, flexible_page)
+      @data = flexible_section_hash
```
Hmm, this is tricky. The parameter name is "what the calling method thinks it's sending you", and the attribute is "what the object thinks of this", so I think it's legit to have them have two names (it's one name for each domain). That said, `@data` could be something like `source_hash`, maybe?

leenagupte replied:
Yeah, I think `source_hash` works. Although to be fair, in the ContentItem model we did keep `content_store_response`: https://github.com/alphagov/frontend/blob/main/app/models/content_item.rb#L11

KludgeKML replied:
I'm going to change data to flexible_section_hash, as per our discussion.

KludgeKML committed:
Add FlexibleSection Factory and base FlexibleSection::Base

- all flexible sections will have a type, the data they were initialised with,
  and a link to the content item (ie flexible page) that contains them.
- factory works similar to the exist landing page block factory:
  https://github.com/alphagov/frontend/blob/7ae68c66b9308f1c45dab8cc2b56ee1814b9b11e/app/models/landing_page/block_factory.rb

    ...but with the simplification that we don't need to provide as much error
    handling, and we don't need a build_all method because there are not
    going to be any nested sections.

KludgeKML commented on app/models/flexible_page/flexible_section_factory.rb:
```diff
@@ -0,0 +1,13 @@
+class FlexiblePage
```
Fixed.

KludgeKML commented on app/models/flexible_page.rb:
```diff
@@ -1,2 +1,9 @@
+  def initialize(content_store_response)
+    super
+
+    @flexible_sections = (content_store_response.dig("details", "flexible_sections") || []).map { |hash| FlexiblePage::FlexibleSectionFactory.build(hash, self) }
```
Test added!

KludgeKML commented on app/helpers/flexible_section_helper.rb:
```diff
@@ -0,0 +1,5 @@
+module FlexibleSectionHelper
+  def render_flexible_section(flexible_section, options: {})
```
Done

leenagupte commented on spec/helpers/flexible_section_helper_spec.rb:
```diff
@@ -0,0 +1,12 @@
+  include described_class
+
+  describe "#render_flexible_section" do
+    it "contains error items" do
```
Minor: I think you could rename this to `it "renders the correct partial for the flexible section"` rather than talking about errors.

leenagupte commented on app/models/flexible_page/flexible_section/base.rb:
```diff
@@ -0,0 +1,11 @@
+  class Base
+    attr_reader :data, :flexible_page, :type
+
+    def initialize(flexible_section_hash, flexible_page)
```
I was getting very confused about what `flexible_page` is here. But it's the content item, right? Perhaps then it should be called `content_item`?

leenagupte replied:
I'm assuming the whole content item is being passed in here so that the `links` hash among others, can still be accessed?

leenagupte commented on app/models/flexible_page/flexible_section_factory.rb:
```diff
@@ -0,0 +1,11 @@
+class FlexiblePage::FlexibleSectionFactory
+  def self.build(flexible_section_hash, flexible_page)
```
Also here, `flexible_page` the full `content_item` right?

leenagupte commented on app/models/flexible_page/flexible_section/rich_content.rb:
```diff
@@ -0,0 +1,19 @@
+      @govspeak = data["govspeak"]
+
+      if data["image"].present?
+        alt, src = data.fetch("image").values_at("alt", "src")
```
I forgot to mention in the previous review, this is a nice use of `fetch` with `values_at`.

leenagupte commented on app/views/flexible_page/show.html.erb:
```diff
@@ -7,6 +7,12 @@
 
+<% content_for :before_content do %>
+  <div class="govuk-width-container">
+    <%= render "govuk_publishing_components/components/contextual_breadcrumbs", content_item: content_item.to_h %>
```
This is much nicer than having to hard-code the breadcrumb 🎉

leenagupte reviewed (commented):
Thank for adding the tests. 

I have a few inline questions about some of the naming. Otherwise it looks good. 🎉

KludgeKML commented on app/views/flexible_page/show.html.erb:
```diff
@@ -7,3 +7,10 @@
 <% end %>
 
+<div class="govuk-width-container">
+  <% @content_item.flexible_sections.each do |flexible_section| %>
```
Done

KludgeKML commented on app/views/flexible_page/show.html.erb:
```diff
@@ -7,3 +7,10 @@
+<div class="govuk-width-container">
+  <% @content_item.flexible_sections.each do |flexible_section| %>
+    <div class="govuk-grid-row">
+      <%= render_flexible_section(flexible_section) %>
```
Solving by removing helper.

leenagupte reviewed (approved):
Thanks for updating the naming 🥇
